File: static/console.js
/**
 * Console functionality for the AI Transcription and Response UI
 */

// Get the console output element
const consoleOutput = document.getElementById('console-output');

// Initialize Socket.IO connection
const socket = io();

/**
 * Add a new message to the console
 * @param {string} message - The message to add
 */
function addConsoleMessage(message) {
    try {
        // Validate input
        if (!message) {
            console.error("Empty message provided to console");
            return;
        }
        
        // Rate limiting - prevent too many messages in short time
        const now = Date.now();
        if (window.lastMessageTime && (now - window.lastMessageTime < 50)) {
            // Queue message for later if too many messages are coming in
            if (!window.messageQueue) window.messageQueue = [];
            window.messageQueue.push(message);
            
            // Process queue after delay if not already scheduled
            if (!window.processingQueue) {
                window.processingQueue = true;
                setTimeout(processMessageQueue, 100);
            }
            return;
        }
        window.lastMessageTime = now;
        
        // Create a new element for the message
        const messageElement = document.createElement('p');
        messageElement.textContent = message;
        messageElement.className = 'console-row';
        
        // Add the message to the console
        consoleOutput.appendChild(messageElement);
        
        // Limit number of messages (prevent memory issues)
        const maxMessages = 1000;
        while (consoleOutput.children.length > maxMessages) {
            consoleOutput.removeChild(consoleOutput.firstChild);
        }
        
        // Auto-scroll to the bottom
        consoleOutput.scrollTop = consoleOutput.scrollHeight;
    } catch (error) {
        console.error("Error adding message to console:", error);
    }
}

/**
 * Process queued messages
 */
function processMessageQueue() {
    if (window.messageQueue && window.messageQueue.length > 0) {
        const message = window.messageQueue.shift();
        addConsoleMessage(message);
        
        if (window.messageQueue.length > 0) {
            setTimeout(processMessageQueue, 50);
        } else {
            window.processingQueue = false;
        }
    } else {
        window.processingQueue = false;
    }
}

/**
 * Clear the console
 */
function clearConsole() {
    try {
        // Remove all messages
        consoleOutput.innerHTML = '';
        
        // Reset message queue if it exists
        if (window.messageQueue) {
            window.messageQueue = [];
            window.processingQueue = false;
        }
    } catch (error) {
        console.error("Error clearing console:", error);
    }
}

// Listen for console messages from the server
socket.on('console_message', function(data) {
    addConsoleMessage(data.message);
});

// Listen for console clear events from the server
socket.on('console_clear', function() {
    clearConsole();
});

// Add event listener for the clear button
document.addEventListener('DOMContentLoaded', function() {
    const clearButton = document.getElementById('clear-console-btn');
    if (clearButton) {
        clearButton.addEventListener('click', function() {
            // Send clear request to server
            fetch('/api/console/clear', {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                }
            })
            .then(response => {
                if (!response.ok) {
                    throw new Error('Network response was not ok');
                }
                // Clear the console locally as well to ensure it's cleared
                clearConsole();
            })
            .catch(error => {
                console.error('Error clearing console:', error);
            });
        });
    }
});

// Export functions for testing
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        addConsoleMessage,
        clearConsole
    };
}

File: console.py
"""
Console module for the AI Transcription and Response UI.
This module provides functionality for managing the console output in the UI.
"""
import datetime
import json
from typing import Dict, Any

class Console:
    """
    Console class for managing console output in the UI.
    Provides methods for adding messages and clearing the console.
    """
    
    def __init__(self):
        """Initialize the Console object."""
        self.messages = []
    
    def printl(self, message: str) -> str:
        """
        Add a new message to the console with timestamp.
        
        Args:
            message: The message to add to the console
            
        Returns:
            The formatted message with timestamp
        """
        # Sanitize input
        if message is None:
            message = ""
        
        # Convert to string if not already
        if not isinstance(message, str):
            try:
                message = str(message)
            except Exception:
                message = "Error: Could not convert message to string"
        
        # Truncate if too long (prevent DoS)
        max_length = 1000
        if len(message) > max_length:
            message = message[:max_length] + "... (truncated)"
        
        # Format the message with timestamp
        timestamp = datetime.datetime.now().strftime("%H:%M:%S")
        formatted_message = f"> [{timestamp}] {message}"
        
        # Store the message in the internal list (limit size to prevent memory issues)
        max_messages = 1000
        if len(self.messages) >= max_messages:
            self.messages = self.messages[-max_messages+1:]
        
        self.messages.append(formatted_message)
        
        return formatted_message
    
    def clear(self) -> None:
        """
        Clear all messages from the console and add a 'Console Cleared' message.
        
        Returns:
            None
        """
        # Clear all messages
        self.messages = []
        
        # Add the "Console Cleared" message
        # This will be displayed after the console is cleared
        self.printl("Console Cleared")
    
    def to_dict(self) -> Dict[str, Any]:
        """
        Convert the console state to a dictionary for JSON serialization.
        
        Returns:
            A dictionary representation of the console state
        """
        return {
            "messages": self.messages
        }
    
    def from_dict(self, data: Dict[str, Any]) -> None:
        """
        Update the console state from a dictionary.
        
        Args:
            data: A dictionary containing console state
        """
        if "messages" in data:
            self.messages = data["messages"]